src/librustc/mir/interpret/value.rs

```diff
@@ -248,8 +258,11 @@ impl<'tcx, Tag> Scalar<Tag> {
         let dl = cx.data_layout();
         match self {
             Scalar::Raw { data, size } => {
-                assert_eq!(size as u64, dl.pointer_size().bytes());
-                Ok(Scalar::Raw { data: dl.signed_offset(data as u64, i)? as u128, size })
+                assert_eq!(u64::from(size), dl.pointer_size.bytes());
+                Ok(Scalar::Raw {
+                    data: u128::from(dl.signed_offset(u64::try_from(data).unwrap(), i)?),
```

Comment:
This `u128` -> `u64` -> `u64` -> `u128` dance deserves a helper function (taking a closure for the `u64` -> `u64`)

Reply:
Speaking of which, there used to be this pattern:
```rust
let x: u128 = ...; // Some potentially large value...
let x_as_u64 = x as u64; // ... that we want to fit in a smaller type
assert_eq!(x_as_u64 as u128, x);
// Use `x_as_u64` now that it's been validated.
```

I believe that `u64::try_from(x).unwrap()` is an appropriate replacement.
So a bunch of `assert_eq!`s can probably be removed now.

Reply:
So you mean something like a `map_u64` method on `Scalar`? Sure.